catalyst_voices/packages/internal/catalyst_voices_services/lib/src/documents/documents_service.dart

```diff
@@ -62,111 +62,188 @@ final class DocumentsServiceImpl implements DocumentsService {
+                              .map((constRef) => constRef.value);
+                        },
+                      )
+                      .expand((element) => element);
+                },
+              )
+              .expand((element) => element.map((e) => e.id)),
+        },
+        onProgress: (value) {
+          if (onProgress == null) {
+            return;
+          }
 
-    final failures = outcomes.whereType<_RefFailure>();
+          /// Each iteration progress is counted equally
+          final prevProgress = completedIterations * progressPerIteration;
+          final curProgress = value * progressPerIteration;
+          final progress = prevProgress + curProgress;
+          onProgress.call(progress);
+        },
+      );
 
-    if (failures.isNotEmpty) {
-      final errors = failures.map((e) => e.failure).toList();
-      throw RefsSyncException(errors);
+      completedIterations++;
+      syncResult += result;
     }
 
     onProgress?.call(1);
 
-    return outcomes.whereType<_RefSuccess>().map((e) => e.success).toList();
+    return syncResult;
   }
 
   @override
   Stream<int> watchCount() {
     return _documentRepository.watchCount();
   }
+
+  Future<DocumentsSyncResult> _sync({
+    required Pool pool,
+    required int batchSize,
+    required DocumentIndexFilters filters,
+    Set<DocumentBaseType> exclude = const {},
+    Set<String> excludeIds = const {},
+    ValueChanged<double>? onProgress,
+  }) async {
+    var page = 0;
+    var remaining = 0;
+    var newDocumentsCount = 0;
+    var failedDocumentsCount = 0;
+
+    onProgress?.call(0);
+
+    do {
+      final index = await _documentRepository.index(
+        page: page,
+        limit: batchSize,
+        filters: filters,
+      );
+
+      final refs = await index.docs
+          .map((e) => e.refs(exclude: exclude))
+          .expand((refs) => refs)
+          .where((ref) => !excludeIds.contains(ref.id))
+          .toSet()
+          .map((ref) {
+            return _documentRepository
+                .isCached(ref: ref)
+                .onError((_, _) => false)
+                .then((value) => value ? null : ref);
+          })
+          .wait
+          .then((refs) => refs.nonNulls.toList());
+
+      final futures = refs.map(
+        (ref) {
+          /// Its possible that missingRefs can be very large
+          /// and executing too many requests at once throws
+          /// net::ERR_INSUFFICIENT_RESOURCES in chrome.
+          /// That's reason for adding pool and limiting max requests.
+          return pool.withResource(() {
+            return _documentRepository
+                .getDocumentData(ref: ref, useCache: false)
+                .then<Result<DocumentData, RefSyncException>>(Success.new)
+                /// Handling errors as Outcome because we have to
+                /// give a change to all refs to finish and keep all info about what
```

Comment:
```suggestion
                /// give a chance to all refs to finish and keep all info about what
```
should this be a chance?